OSSM-3796: Refactor adding services to a mesh (#59177)
openshift/openshift-docs: modules/ossm-about-adding-namespace.adoc

// Module included in the following assemblies:
//
// * service_mesh/v2x/installing-ossm.adoc

:_content-type: CONCEPT
[id="ossm-about-adding-namespace_{context}"]
= About adding projects to a service mesh

A Kubernetes namespace provides a mechanism to scope resources in a cluster. In {product-title}, a project is essentially a Kubernetes namespace with additional annotations. After you create the `ServiceMeshControlPlane` resource, you add projects to the mesh. You can add projects to an existing {SMProductShortName} resource using the web console or the CLI.

There are two methods to add a project to a service mesh:

* Specify the project in the `ServiceMeshMemberRoll` resource.

* Create the `ServiceMeshMember` resource in the project.

> == ServiceMeshMemberRoll method

Author: Done

This is the simplest method to add a project to a service mesh. Specify the project in the `spec.members` field of the `ServiceMeshMemberRoll` resource.   

Adding projects using this methods requires a user to have mesh admin privileges.

> == ServiceMeshMember method

Reviewer: ```suggestion
.ServiceMeshMember method
```

Author: Done

// Module included in the following assemblies:
//
// * service_mesh/v2x/installing-ossm.adoc

:_content-type: CONCEPT
[id="ossm-about-adding-namespace_{context}"]
= About adding projects to a service mesh

> A project contains services; however, the services are not available to the mesh unless you add the project to the service mesh.

Reviewer: Optional: (Positive tone)
```suggestion
A project contains services; however, the services are only available to the mesh if you add the project to the service mesh.
```

[NOTE]
====

> In OpenShift Container Platform, a project is essentially a Kubernetes namespace with additional annotations, such as the range of user IDs that can be used in the project. Typically, the {product-title} web console uses the term project, and the CLI uses the term namespace, but the terms are essentially synonymous.

Author: Done

====

You can add projects to an existing service mesh using either the web console or the CLI. There are two methods to add a project to a mesh:

* Specifying a project in the `ServiceMeshMemberRoll` resource.

> * Creating a `ServiceMeshMember` resource in a project.

Author: Done

.ServiceMeshMemberRoll method

> This is the simplest way to add a project to a service mesh. To add a project, specify the project namespace where your applications reside in the `spec.members` field of the `ServiceMeshMemberRoll` resource.

Author: Done.

[NOTE]
====

> Adding projects using this method requires a user to have mesh admin privileges.

Reviewer 2: ```suggestion
Adding projects using this method requires the user to have the `update servicemeshmemberrolls` privilege and the `update pods` privilege in the project that is being added.
```

Author: Done

====

> The `ServiceMeshMemberRoll` resource specifies which projects are controlled by the the `ServiceMeshControlPlane` resource. The control plane includes Istiod, Ingress and Egress Gateways, and other components, such as Kiali and Jaeger. The control plane must be deployed in a separate namespace than the {SMProductShortName} Operators and the data plane applications and services.

Author: Done. Thank you for noticing that!

> The `ServiceMeshMemberRoll` resource lives alongside the `ServiceMeshControlPlane` resource, and specifies the list of projects that should be controlled by the control plane. The control plane, which includes Istiod, Ingress and Egress Gateways, and additional components such as Kiali and Jaeger, must be deployed in a different namespace than the {SMProductShortName} Operators and the applications and services (which are part of the data plane).

Reviewer: Maybe reword it for clarity:
```suggestion
A `ServiceMeshMemberRoll` resource specifies which projects are controlled by a `ServiceMeshControlPlane` resource. The control plane has Istiod, Ingress, and Egress Gateways, and other components like Kiali and Jaeger. The control plane must be in a separate namespace from the {SMProductShortName} Operators and the data plane applications and services.
```
This suggestion probably needs SME review.

Author: Done

> * If you already have an application, workload, or service to add to a `ServiceMeshMemberRoll` resource, see the instructions for adding or removing projects from the service mesh using the `ServiceMeshMemberRoll` resource with the xref:../../service_mesh/v2x/ossm-create-mesh.adoc#ossm-add-project-member-roll-recourse-console_ossm-create-mesh[web console] or with the xref:../../service_mesh/v2x/ossm-create-mesh.adoc#ossm-add-project-member-roll-resource-cli_ossm-create-mesh[CLI]. 

Reviewer 2: ```suggestion
* If you already have an application, workload, or service to add to the service mesh, see the instructions for adding or removing projects from the service mesh using the `ServiceMeshMemberRoll` resource with the xref:../../service_mesh/v2x/ossm-create-mesh.adoc#ossm-add-project-member-roll-recourse-console_ossm-create-mesh[web console] or with the xref:../../service_mesh/v2x/ossm-create-mesh.adoc#ossm-add-project-member-roll-resource-cli_ossm-create-mesh[CLI]. 
```

> * Alternatively, to install a sample application called Bookinfo and add it to a `ServiceMeshMemberRoll` resource, skip to the tutorial for installing the xref:../../service_mesh/v2x/ossm-create-mesh.adoc#ossm-tutorial-bookinfo-overview_ossm-create-mesh[Bookinfo example application] to see how an application works in {SMProductName}.

Reviewer 2: BTW, every time I read this, I read it as "skip the tutorial" instead of "skip *to* the tutorial". You might want to rephrase.

Reviewer 3: +1

Reviewer: ```suggestion
* Alternatively, to install a sample application called Bookinfo and add it to a `ServiceMeshMemberRoll` resource, see the xref:../../service_mesh/v2x/ossm-create-mesh.adoc#ossm-tutorial-bookinfo-overview_ossm-create-mesh[Bookinfo example application] tutorial.
```

.ServiceMeshMember method

> A `ServiceMeshMember` resource provides a way to add a project to a service mesh without modifying the `ServiceMeshMemberRoll` resource. To add a project, create a `ServiceMeshMember` resource in the project that you want to add to the service mesh. The namespace that contains the `ServiceMeshMember` resource becomes part of the `ServiceMeshMemberRoll` resource, and the services that reside in the namespace are made available to the mesh. For more information, see the instructions for adding projects to the service mesh using the `ServiceMeshMember` resource with the xref:../../service_mesh/v2x/ossm-create-mesh.adoc#ossm-adding-project-using-smm-resource-console_ossm-create-mesh[web console] or with the xref:../../service_mesh/v2x/ossm-create-mesh.adoc#ossm-adding-project-using-smm-resource-cli_ossm-create-mesh[CLI].

Reviewer 2: I can't find the proper wording, but this needs to be changed so that it's clear that the user is creating the resource in the project that they are adding to the mesh. Something like:
```suggestion
A `ServiceMeshMember` resource provides a way to add a project to a service mesh without modifying the `ServiceMeshMemberRoll` resource. To add a project, create a `ServiceMeshMember` resource in that same project. When the `ServiceMeshMember` object is processed by the {SMProductShortName} Operator, the project appears in the `status.members` list in the `ServiceMeshMemberRoll` resource, and the services that reside in the project are made available to the mesh. For more information, see the instructions for adding projects to the service mesh using the `ServiceMeshMember` resource with the xref:../../service_mesh/v2x/ossm-create-mesh.adoc#ossm-adding-project-using-smm-resource-console_ossm-create-mesh[web console] or with the xref:../../service_mesh/v2x/ossm-create-mesh.adoc#ossm-adding-project-using-smm-resource-cli_ossm-create-mesh[CLI].
```
